In [12]:
import pandas as pd
import numpy as np
from scipy.stats import entropy as en

In [15]:
def calcula_entropia_maxima(num_eventos):
    prob_uniforme = 1 / num_eventos
    entropia_maxima = -num_eventos * prob_uniforme * np.log2(prob_uniforme)
    return entropia_maxima

In [14]:
data = pd.read_csv('beer_data.csv', low_memory=False)
#cleaning the null values
data = data.dropna(how='any',axis=0)
data.head()

name  \
0                         Vanilla Cream Ale   
1        Avg. Perfect Northeast IPA (NEIPA)   
2              Sierra Nevada Pale Ale Clone   
3             Zombie Dust Clone - ALL GRAIN   
4  Russian River Pliny the Elder (original)   

                                                 url     method  \
0       /homebrew/recipe/view/1633/vanilla-cream-ale  All Grain   
1  /homebrew/recipe/view/363082/avg-perfect-north...  All Grain   
2  /homebrew/recipe/view/28546/sierra-nevada-pale...  All Grain   
3  /homebrew/recipe/view/5916/zombie-dust-clone-a...  All Grain   
4  /homebrew/recipe/view/37534/russian-river-plin...  All Grain   

                            style  batch     og     fg   abv     ibu  color  \
0                       Cream Ale   21.8  1.055  1.013  5.48   19.44   4.83   
1  Specialty IPA: New England IPA   21.8  1.062  1.013  6.50   59.26   5.20   
2               American Pale Ale   24.6  1.055  1.013  5.58   39.79   8.00   
3                    American IPA   22.7  1.061  1.016  5.94   62.42   8.50   
4                    Imperial IPA   22.7  1.072  1.018  7.09  232.89   6.33   

   ...  hops_process hops_process_time hops_IBU  hops_bill  \
0  ...          Boil            60 min    11.42       33.3   
1  ...          Boil            10 min    13.55        7.3   
2  ...          Boil            60 min    22.62        8.3   
3  ...    First Wort             0 min    15.57        8.6   
4  ...          Boil            90 min   171.54       28.0   

                                      yeast_name  yeast_attenuation  \
0                           Wyeast - Kölsch 2565                76%   
1                   Wyeast - London Ale III 1318                78%   
2  Fermentis - Safale - American Ale Yeast US-05                76%   
3    Fermentis - Safale - English Ale Yeast S-04                75%   
4                     Wyeast - American Ale 1056                75%   

   yeast_flocculation yeast_temp_min  yeast_temp_max  yeast_starter  
0                 Low           56.0            70.0            Yes  
1                High           64.0            74.0             No  
2              Medium           54.0            77.0            Yes  
3                High           54.0            77.0            Yes  
4             Med-Low           60.0            72.0             No  

[5 rows x 34 columns]

In [28]:
xData = data.groupby(['style']).size().reset_index(name='quantidade')
xData = xData[~xData['style'].isin(['--'])]
xData.tail()

style  quantidade
174             Wheatwine          96
175   Wild Specialty Beer          99
176  Winter Seasonal Beer         340
177               Witbier        2071
178        Wood-Aged Beer         105

In [29]:
total = xData.sum()
print(total.quantidade)

134450


In [33]:
prob = []
for i in xData.quantidade:
    prob.append(i/total.quantidade)
prob

[0.0025065079955373746,
 0.0005578281889178133,
 0.0002528821123094087,
 0.02554109334325028,
 0.003919672740795835,
 0.013172182967645964,
 0.15539605801413164,
 0.004194867980661956,
 0.028389735961323912,
 0.10394942357753811,
 0.012034213462253626,
 0.016727407958348827,
 0.003250278914094459,
 0.0114540721457791,
 0.00458162885831164,
 7.437709185570844e-06,
 0.0011305317962067682,
 0.0014057270360728895,
 0.004261807363332094,
 0.0071178876905912975,
 0.007021197471178877,
 0.006671625139457047,
 0.004819635552249907,
 0.008486426180736334,
 0.0038378579397545555,
 0.007802156935663816,
 0.006054295277054667,
 0.006277426552621792,
 0.0023280029750836744,
 0.02748977314986984,
 0.0041651171439196725,
 0.000996653030866493,
 0.0010189661584232056,
 0.006061732986240238,
 0.004596504276682782,
 0.002945332837486054,
 0.005630345853477129,
 7.437709185570845e-05,
 0.002342878393454816,
 0.007244328746746002,
 7.437709185570844e-06,
 0.002536258832279658,
 0.000431387132763109,
 0.00

In [36]:
entropiaCervejas = en(prob)
surpresaEsperada = 1 / entropiaCervejas
entropiaMaxima = calcula_entropia_maxima(int(total.quantidade))


print(f"A entropia dos estilos de cervejas é: {entropiaCervejas}")
print(f"A surpresa esperada ao retirar uma cerveja é: {surpresaEsperada}")
print(f"A entropia máxima dos estilos de cervejas é: {entropiaMaxima}")

A entropia dos estilos de cervejas é: 4.084564677981239
A surpresa esperada ao retirar uma cerveja é: 0.24482413153860047
A entropia máxima dos estilos de cervejas é: 17.036710229647596


# Conclusão

*Com base nos resultados apresentados sobre a entropia dos estilos de cervejas, podemos inferir importantes características desse universo cervejeiro. A entropia, medida de incerteza e desordem, revela um índice de 4.08, indicando uma quantidade moderada de variedade e diversidade entre os estilos. Essa pontuação sugere uma distribuição equilibrada de frequências entre as diferentes categorias de cervejas. A surpresa esperada ao retirar uma cerveja, correlacionada à entropia, é relativamente baixa (0.24), indicando que, em média, a escolha de uma cerveja não é altamente imprevisível. Os estilos de cervejas parecem seguir um padrão previsível com base na distribuição de entropia. Contrapondo-se a essa previsibilidade, a entropia máxima, alcançando 17.04, sugere um amplo espectro de estilos de cervejas com uma distribuição altamente diversificada. No entanto, a entropia real observada (4.08) está substancialmente aquém do máximo possível, indicando que a variedade de estilos não atinge seu potencial máximo. Em resumo, os resultados proporcionam insights valiosos sobre a diversidade e previsibilidade dos estilos de cervejas, onde a entropia moderada sugere uma variedade razoável e a surpresa esperada baixa indica certo grau de previsibilidade nas escolhas. A comparação com a entropia máxima destaca que, embora haja diversidade, ela ainda não atinge o ápice possível.*

Referência Dataset: https://www.kaggle.com/datasets/basaltier/homebrew-beer-data